src/storage/data_table.cpp

```diff
@@ -29,31 +46,50 @@ void DataTable::Select(transaction::TransactionContext *const txn, const TupleSl
     // Here we will need to check that the version pointer did not change during our read. If it did, the content
     // we have read might have been rolled back and an abort has already unlinked the associated undo-record,
     // we will have to loop around to avoid a dirty read.
+    visible = Visible(slot, accessor_);
   } while (version_ptr != AtomicallyReadVersionPtr(slot, accessor_));
 
   // Nullptr in version chain means no version visible to any transaction alive at this point.
   // Alternatively, if the current transaction holds the write lock, it should be able to read its own updates.
-  if (version_ptr == nullptr || version_ptr->Timestamp().load() == txn->TxnId().load()) return;
+  if (version_ptr == nullptr || version_ptr->Timestamp().load() == txn->TxnId().load()) {
+    return visible;
+  }
 
   // Apply deltas until we reconstruct a version safe for us to read
   // If the version chain becomes null, this tuple does not exist for this version, and the last delta
   // record would be an undo for insert that sets the primary key to null, which is intended behavior.
   while (version_ptr != nullptr &&
          transaction::TransactionUtil::NewerThan(version_ptr->Timestamp().load(), txn->StartTime())) {
-    StorageUtil::ApplyDelta(accessor_.GetBlockLayout(), *(version_ptr->Delta()), out_buffer);
+    auto modifies_logical_delete_column = StorageUtil::DeltaModifiesLogicalDelete(*(version_ptr->Delta()));
+    if (modifies_logical_delete_column == LogicalDeleteModificationType::NONE) {
+      // Normal delta to be applied. Does not modify the logical delete column.
+      StorageUtil::ApplyDelta(accessor_.GetBlockLayout(), *(version_ptr->Delta()), out_buffer);
+    } else if (modifies_logical_delete_column == LogicalDeleteModificationType::INSERT) {
+      // Applying the undo of an INSERT makes the tuple invisible to this txn.
+      visible = false;
```

Comment:
I am not convinced we need to special case inserts and deletes. We still end up with the correct buffer, just need to check the resulting buffer at the end for visibility. Also, we don't really do extra work because as you said most likely inserts and deletes are at the end and beginning of a version chain, respectively.

Reply:
The idea is that the output buffer doesn't have a column for the logical delete column, so we're not actually applying deltas in that case. The notion of the logical delete column shouldn't be exposed above this layer, so the ProjectedRow won't hold an attribute for it.